### DISCUSSION TITLE
Reconsider use of "dependency" terminology

### DISCUSSION THREAD

**Discussion by billcrook** [fastapi/fastapi#4023] [Questions]
Dependency injection means a very different thing to the software community writ large. This creates a cognitive dissonance between the generally understood definition of dependency injection and the feature in fastapi. This can be seen in the variety of issues such as https://github.com/tiangolo/fastapi/issues/1514 and https://github.com/tiangolo/fastapi/issues/504. Anyone coming from a mature DI/IOC framework (lots of experienced devs out there) is faced with reconciling these differences. What is implemented in fastapi is more akin to a request/path transformer functions that can be chained. If there is no plan to implement a full featured DI/IOC module which provides basic features such as scopes, then rewording these "dependency" concepts in fastapi might reduce this dissonance.

**Comment by adriangb** [comment]
Hey @billcrook, could you take a look at #3641? I attempt to lay out a framework for actually implementing full DI in FastAPI. I've also been prototyping an implementation in https://github.com/adriangb/di and https://github.com/adriangb/starlette-di

**Reply by billcrook** [reply]
@adriangb I see no comments from the [benevolent dictator](http://oss-watch.ac.uk/resources/governancemodels) on this proposal and hence no reason to believe it will be integrated into FastAPI. I applaud your initiative to build a brand new DI framework, but if something _was_ to be added to FastAPI, why not integrate an established DI framework such as https://github.com/ets-labs/python-dependency-injector? FastAPI is founded on the concept of bringing multiple frameworks together (in fact that's all it really does + documentation) and I see no reason why something like python-dependency-injector couldn't be folded under FastAPI. IMO, either rephrase the terminology or bring in a mature DI/IOC system. OFC, this is just my opinion and I understand others may not have the same view.

tagging @rmk135 for awareness

**Reply by adriangb** [reply]
Yes, you're right, I have not gotten any input from Sebastian, and there is currently no reason to believe it will be integrated into FastAPI.

The main motivation for making the new DI framework was to expand the current FastAPI API in a backwards compatible manner, while also making sure that it can serve the purposes FastAPI needs (like introspecting dependencies to build the docs). I'd hope this makes it easier to incorporate, either directly or as inspiration / a proving ground.

I do think bringing in an established framework generally makes sense, and python-dependency-injector is probably the most established framework out there. I'm just not sure the API can be made compatible with FastAPI's current API, although I'd love to be proven wrong.

**Reply by billcrook** [reply]
FYI, I experimented with this today: https://gist.github.com/billcrook/88733370bff2bb86136dc852476109d4

**Reply by adriangb** [reply]
That's certainly an interesting approach.
I wonder if it could be made a bit more in line with the current API if the FastAPI App created the Container, ran the wiring (perhaps when `include_router` is called) and such? That should be doable by subclassing `fastapi.applications.FastAPI` and `fastapi.routing.APIRouter`.

I'd also be interested to see how this approach would tackle building docs (introspection is currently done via the DI system) and other nuances.

**Reply by billcrook** [reply]
> That's certainly an interesting approach.

Haha, that is an understatement :) 

It was just an experiment/hack to see if I could wire them together somehow. In no way am I recommending this as a solution. There might be some cleaner way to do this using [APIRouter.dependency_overrides_provider](https://github.com/tiangolo/fastapi/blob/b8c4149e89d1c97d204ac3f965e6144e3dc126a9/fastapi/routing.py#L442) but it is not documented and I haven't had a chance to investigate.

**Reply by adriangb** [reply]
`dependency_overrides_provider` is a reference to an object w/ a `dependency_overrides` attribute which is a mapping of `callable in signature => replacement callable`. In practice, this is referring to [`FastAPI.dependency_overrides`](https://github.com/tiangolo/fastapi/blob/864643ef7608d28ac4ed321835a7fb4abe3dfc13/fastapi/applications.py#L124). The logic for that is [here](https://github.com/tiangolo/fastapi/blob/864643ef7608d28ac4ed321835a7fb4abe3dfc13/fastapi/dependencies/utils.py#L479-L494). So it's really not much different than your `app.dependency_overrides = DictProxy` approach.

**Reply by billcrook** [reply]
Yeah, and it looks like I would need the DictProxy anyway to keep FastAPI happy about the callable signature.

**Reply by billcrook** [reply]
FYI, my [experiment](https://gist.github.com/billcrook/88733370bff2bb86136dc852476109d4) doesn't actually hook in any deeper than already outlined in the [fastapi example in the python-dependency-injector docs](https://python-dependency-injector.ets-labs.org/examples/fastapi-sqlalchemy.html). There are a bunch of [FastAPI tickets over in that project](https://github.com/ets-labs/python-dependency-injector/issues?q=is%3Aissue+is%3Aopen+fastapi). If ets-labs/python-dependency-injector/issues/363 is tackled and api docs are figured out, then maybe python-dependency-injector could be the DI/IOC container for FastAPI.

**Comment by billcrook** [comment]
In case anyone comes across this discussion, this alternative to fastapi looks interesting: https://xpresso-api.dev/0.38.1/tutorial/dependencies/scopes/

**Reply by billcrook** [reply]
@adriangb 👍

**Reply by adriangb** [reply]
I think it's worth pointing out that Xpresso is under active development and you shouldn't use it in production unless you understand that things may break. FastAPI is more stable and just generally is a much more established project. So I don't know that it's an _alternative_ other than it being a similar codebase offering similar features.

**Comment by xkortex** [comment]
> Dependency injection means a very different thing to the software community writ large. 

Could you elaborate why `Depends` is "very different" from established DI systems in general? Does `python-dependency-injector` meet your criteria of "mature DI/IOC framework"? I'm fairly new to the DI pattern, with FastAPI being my first real exposure to using it, so I want to make sure I'm not "learning any bad habits" so to speak, if `Depends` is not a good mental model to extrapolate from. I've looked at `python-dependency-injector` and it seems to me that `Depends` is still a form of dependency injection, it's just nerfed and limited in scope. Personally I see it as a lower rung of sophistication, but still on the ladder of DI patterns, rather than "very different", but maybe that is just a word choice thing. 

> I do think bringing in an established framework generally makes sense, and python-dependency-injector is probably the most established framework out there. 

I love FastAPI's `Depends` system, but my one frustration with it is that it does not generalize outside of FastAPI, such as for Celery tasks. I have to use mocks with my celery tasks rather than the fixtures I use with endpoints, and I gotta say, the fixtures with DI and `dependency_overrides` is far more ergonomic.  

I spent some time tooling around with the `solve_dependencies` function to see if I could leverage it for my own purposes and create job queue task decorators with the same ergonomics of FastAPI `APIRoute`s. The quick and dirty proof of concept I came up with is transform the incoming message in a `starlette.requests.Request`. It feels like `solve_dependencies` is a tad too opinionated about the input being a `Request` object (which makes complete sense - that's what it's designed to do), but I wonder if the dependency injection part of the machinery could be separated from the request parsing part, and operate on more generic Containers. That would (in theory) then make it easier to integrate with `dependency-injectors`, without exposing neophytes to the heavy DI stuff; thus FastAPI would maintain the same outward Depends interface and low barrier of entry. 

I think the most important concern is maintaining that low barrier of entry to FastAPI's DI system, one that "just works" without any extra wiring, even if that means diverging from more canonical DI patterns. If that shiny veneer can be maintained on the surface, it's less important how that DI happens under the hood. But it is also an important as a secondary concern to make sure developers who are introduced to DI through FastAPI don't ingrain any "bad habits".

**Reply by adriangb** [reply]
It can be separated out: https://github.com/adriangb/di
